base_layer/mmr/src/change_tracker.rs

```diff
@@ -0,0 +1,251 @@
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    backend::ArrayLike,
+    error::MerkleMountainRangeError,
+    pruned_mmr::{prune_mutable_mmr, PrunedMutableMmr},
+    Hash,
+    MutableMmr,
+};
+use croaring::Bitmap;
+use digest::Digest;
+use std::{mem, ops::Deref};
+
+/// A struct that wraps an MMR to keep track of changes to the MMR over time. This enables one to roll
+/// back changes to a point in history. Think of `MerkleChangeTracker` as 'git' for MMRs.
+///
+/// [MutableMMr] implements [std::ops::Deref], so that once you've wrapped the MMR, all the immutable methods are
+/// available through the auto-dereferencing.
+///
+/// The basic philosophy of `MerkleChangeTracker` is as follows:
+/// * Start with a 'base' MMR. For efficiency, you usually want to make this a [pruned_mmr::PrunedMmr], but it
+/// doesn't have to be.
+/// * We then maintain a change-list for every append and delete that is made on the MMR.
+/// * You can `commit` the change-set at any time, which will create a new [MerkleCheckPoint] summarising the
+/// changes, and the current change-set is reset.
+/// * You can `rewind` to a previously committed checkpoint, p. This entails resetting the MMR to the base state and
+/// then replaying every checkpoint in sequence until checkpoint p is reached. `rewind_to_start` and `replay` perform
+/// similar functions.
+/// * You can `reset` the ChangeTracker, which clears the current change-set and moves you back to the most recent
+/// checkpoint ('HEAD')
+pub struct MerkleChangeTracker<D, BaseBackend, CpBackend>
+where
+    D: Digest,
+    BaseBackend: ArrayLike<Value = Hash>,
+{
+    base: MutableMmr<D, BaseBackend>,
+    mmr: PrunedMutableMmr<D>,
+    checkpoints: CpBackend,
+    // The hashes added since the last commit
+    current_hashes: Vec<Hash>,
```

Comment:
```suggestion
    current_additions: Vec<Hash>,
```
Not entirely sure if these are only additions, but it seems to be

Reply:
Changed